site/src/xServices/workspaceSchedule/workspaceScheduleXService.ts

```diff
@@ -50,160 +53,227 @@ export type WorkspaceScheduleEvent =
+        },
+        clearGetWorkspaceError: (context) => {
+          assign({ ...context, getWorkspaceError: undefined })
+        },
+        // action instead of service because we fire and forget so that the
+        // user can return to the workspace page to see the restart
+        restartWorkspace: (context) => {
+          if (context.workspace && context.template) {
+            return API.startWorkspace(
+              context.workspace.id,
+              context.template.active_version_id,
+            )
+          }
+        },
       },
-    },
 
-    services: {
-      getWorkspace: async (_, event) => {
-        return await API.getWorkspaceByOwnerAndName(
-          event.username,
-          event.workspaceName,
-        )
-      },
-      checkPermissions: async (context) => {
-        if (context.workspace) {
-          return await API.checkAuthorization({
-            checks: permissionsToCheck(context.workspace),
-          })
-        } else {
-          throw Error(
-            "Cannot check permissions without both workspace and user id",
+      services: {
+        getWorkspace: async (_, event) => {
+          return await API.getWorkspaceByOwnerAndName(
+            event.username,
+            event.workspaceName,
           )
-        }
-      },
-      submitSchedule: async (context, event) => {
-        if (!context.workspace?.id) {
-          // This state is theoretically impossible, but helps TS
-          throw new Error("Failed to load workspace.")
-        }
+        },
+        getTemplate: async (context) => {
+          if (context.workspace) {
+            return await API.getTemplate(context.workspace.template_id)
+          } else {
+            throw Error("Can't fetch template without workspace.")
+          }
+        },
+        checkPermissions: async (context) => {
+          if (context.workspace) {
+            return await API.checkAuthorization({
+              checks: permissionsToCheck(context.workspace),
+            })
+          } else {
+            throw Error(
+              "Cannot check permissions without both workspace and user id",
+            )
+          }
+        },
+        submitSchedule: async (context, event) => {
+          if (!context.workspace?.id) {
+            // This state is theoretically impossible, but helps TS
+            throw new Error("Failed to load workspace.")
+          }
 
-        if (event.autoStartChanged) {
-          await API.putWorkspaceAutostart(context.workspace.id, event.autoStart)
-        }
-        if (event.autoStopChanged) {
-          await API.putWorkspaceAutostop(context.workspace.id, event.ttl)
-        }
+          if (event.autoStartChanged) {
+            await API.putWorkspaceAutostart(
+              context.workspace.id,
+              event.autoStart,
+            )
+          }
+          if (event.autoStopChanged) {
+            await API.putWorkspaceAutostop(context.workspace.id, event.ttl)
```

Comment:
What happens if this errors? I don't see a catch, so I'm not sure how we'd enter the error state on line 157 of this machine. I would want to avoid prompting users to restart their workspace if no change was actually made.

Reply:
I wasn't sure if it was a problem that I'm not returning the result from both promises here, so I tried hard-coding each api call to return a rejected promise, and it behaved the way I intended: if one api call errors and that part of the schedule is not changed, there's no error. If the error api call is called, you stay on the form and get an error message. If you change the start but not the stop, you don't see the prompt.